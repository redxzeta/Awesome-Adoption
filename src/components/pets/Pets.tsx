import PetCardFlex, { PawHubContainer } from "components/layout/Grid/PetCardFlex";
import { Card } from "react-daisyui";
import { Link } from "react-router-dom";
import useSWR from "swr";
import { PetListType } from "types/PetType";

import { usePetAuth } from "../../context/TokenContext";
import { fetcher } from "../../utils/petInfoFetcher";
import Bird from "./bird.jpg";
import Cat from "./cat.jpg";
import Dog from "./doggo.jpg";
import Horse from "./horse.jpg";
import Placeholder from "./placeholder-light.png";
import Rabbit from "./rabbit.jpg";

type PetLink = {
  type: PetListType;
  img: string;
};

const linkData: PetLink[] = [
  {
    img: Dog,
    type: "Dog"
  },
  {
    img: Cat,
    type: "Cat"
  },
  {
    img: Bird,
    type: "Bird"
  },
  {
    img: Horse,
    type: "Horse"
  },
  {
    img: Rabbit,
    type: "Rabbit"
  }
];

export default function Pets() {
  return (
    <PawHubContainer>
      <h1 className="text-5xl font-bold font-amatic mb-10">Adopt Your Buddy</h1>

      <PetCardFlex>
        {linkData.map(pet => (
          <AnimalType img={pet.img} type={pet.type} link={`${pet.type.toLowerCase()}`} key={pet.type} />
        ))}
        <RandomPet />
      </PetCardFlex>
    </PawHubContainer>
  );
}

export const AnimalType = ({ type, img, link }: { type: string; img: string; link: string }) => {
  let linkDataTypes: string[] = [];
  linkData.forEach(({ type }) => {
    linkDataTypes = [...linkDataTypes, type];
  });
  const isLinkDataType = linkDataTypes.includes(type);

  return (
    <article className="lg:px-4 px-1 w-full md:1/3 lg:w-1/4 my-4 fadeInUp ">
      <Card imageFull>
        <Card.Image src={img} alt={type} className="w-full max-h-80 " />
        <Card.Body>
          <Card.Title tag="h2">
            {type} {isLinkDataType ? "Buddies" : "Buddy"}
          </Card.Title>
          <p>If a {type} chews shoes whose shoes does he choose?</p>
          <Card.Actions className="justify-end">
            <Link to={link} className="btn btn-primary">
              Click Here
            </Link>
          </Card.Actions>
        </Card.Body>
      </Card>
    </article>
  );
};

const RandomPet = () => {
  const randomPetURL = "https://api.petfinder.com/v2/animals?limit=1&sort=random";
  const { tokenHeaders } = usePetAuth();

  const { error, data } = useSWR(tokenHeaders ? [randomPetURL, tokenHeaders] : null, fetcher, {
    revalidateOnFocus: false,
    revalidateIfStale: false,
    revalidateOnReconnect: false
  });
  const isLoading = !error && !data;

  if (isLoading)
    return (
      <article className="lg:px-4 px-1 w-full md:1/3 lg:w-1/4 my-4 fadeInUp" role={"status"}>
        <Card imageFull>
          <div className="object-cover w-full animate-pulse  bg-primary" />
          <Card.Body>
            <Card.Title tag="h2">
              <span className="visually-hidden">Loading...</span>
            </Card.Title>

            <Card.Actions className="justify-end"></Card.Actions>
          </Card.Body>
        </Card>
      </article>
    );
  if (error || !data) {
    return <h1>An Error Occurred</h1>;
  }

  const randomPetImage = data.photos[0].medium ?? Placeholder;

  // assigns randomPetImage according to if a photo is available
  return <AnimalType type={data.name} img={randomPetImage} link={`/animal/${data.id}`} />;
};
