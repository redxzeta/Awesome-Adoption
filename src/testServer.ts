import { createClient } from "@supabase/supabase-js";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { ProfileType } from "utils/supaFetcher";

import PlaceImage from "./components/about/__tests__/placeholder50px.png";
import DogSample from "./testData/sample.json";

const contributors = [
  {
    id: 1,
    login: "abe",
    html_url: "somesite.com",
    avatar_url: PlaceImage
  },
  {
    id: 2,
    login: "label",
    html_url: "asomesite.com",
    avatar_url: PlaceImage
  }
];

const petList = {
  animal: {
    id: 1,
    url: "https://wwww.mandalorian.com",
    name: "Baby Yoda",
    type: "Grogu",
    contact: {
      email: "yoda@onefor.me"
    },
    breeds: {
      primary: "Grogu"
    },
    colors: {
      primary: "green"
    },
    age: "Baby",
    gender: "male",
    description:
      "He is a toddler member of the same unnamed alien species as the Star Wars characters Yoda and Yaddle, with whom he shares a strong ability in the Force",
    photos: [
      {
        medium: "babyYoda.medium.jpg",
        large: "babyYoda.large.jpg"
      }
    ]
  }
};

const petListFav = [
  {
    animal: {
      id: 2,

      name: "Courage the Cowardly Dog",
      type: "dog",
      age: "Baby",
      gender: "male",
      description: "Woof",
      primary_photo_cropped: {
        medium: "courage.jpg"
      },
      breeds: {
        primary: "dog"
      },
      photos: [
        {
          medium: "courage.medium.jpg",
          large: "courage.large.jpg"
        }
      ]
    }
  },
  {
    animal: {
      id: 3,
      name: "jake",
      type: "dog",
      age: "young",
      gender: "male",
      description: "finns friend",
      primary_photo_cropped: {
        medium: "jake.png"
      },
      breeds: {
        primary: "dog"
      },
      photos: [
        {
          medium: "finn.medium.jpg",
          large: "finn.large.jpg"
        }
      ]
    }
  },
  {
    animal: {
      id: 4,
      name: "puss n boots",
      type: "cat",
      age: "young",
      gender: "male",
      description: "masterr of sowrds",
      primary_photo_cropped: {
        medium: "boots.jpg"
      },
      breeds: {
        primary: "cat"
      },
      photos: [
        {
          medium: "cat.medium.jpg",
          large: "cat.large.jpg"
        }
      ]
    }
  }
];

const fakenewProfiles: ProfileType[] = [
  {
    id: "35",
    username: "supaAwesome",
    description: "Supa Awesome yet ",
    avatar_url: "cuteDoggoAndCat.jpg",
    favoritepets: [],
    background: {
      id: 1,
      background_url: "someImage.png"
    }
  },
  {
    id: "36",
    username: "supaPet",
    description: "Supa Awesome yet ",
    avatar_url: "cuteDoggoAndCat.jpg",
    favoritepets: [
      { id: 1, pet: "2", created_at: new Date("01/01/2020") },
      { id: 2, pet: "3", created_at: new Date("01/02/2020") }
    ],
    background: {
      id: 1,
      background_url: "someImage.png"
    }
  }
];

const server = setupServer(
  http.get("https://api.github.com/repos/redxzeta/Awesome-Adoption/contributors", () => {
    return HttpResponse.json(contributors, { status: 200 });
  }),
  http.post("https://api.petfinder.com/v2/oauth2/token", () => {
    return HttpResponse.json({ access_token: "234566" }, { status: 200 });
  }),
  http.get("https://api.petfinder.com/v2/animals/:id", info => {
    const { id } = info.params;

    switch (id) {
      case "1":
        return HttpResponse.json(petList, { status: 200 });
      case "2":
        return HttpResponse.json(petListFav[0], { status: 200 });
      case "3":
        return HttpResponse.json(petListFav[1], { status: 200 });
      case "4":
        return HttpResponse.json(petListFav[2], { status: 200 });
      default:
        return HttpResponse.json({ message: "Yoda does not exist" }, { status: 404 });
    }
  }),
  http.get("https://api.petfinder.com/v2/animals", ({ request }) => {
    const url = new URL(request.url);
    const sort = url.searchParams.get("sort");
    const limit = url.searchParams.get("limit");
    const type = url.searchParams.get("type");
    const location = url.searchParams.get("location");
    // const page = info.url.searchParams.get("page");
    if (sort === "random" && limit === "1") {
      return HttpResponse.json(petList, { status: 200 });
    } else if (sort === "random" && limit === "3") {
      const threePets = DogSample.animals.slice(0, 3);
      return HttpResponse.json({ animals: threePets }, { status: 200 });
    } else if (type === "dog" && location && limit === "12") {
      return HttpResponse.json(DogSample, { status: 200 });
    }
    return HttpResponse.json(contributors, { status: 200 });
  }),

  // Login
  http.post("https://test.supabase.co/auth/v1/token", () => {
    return HttpResponse.json(
      {
        access_token: "fake_access_token",
        expires_at: 9999999999,
        expires_in: 3600,
        refresh_token: "uCtyqUsj5FJqtIxCkz2Mvg",
        token_type: "bearer",
        user: { id: "1234" }
      },
      { status: 200 }
    );
  }),
  http.post("https://test.supabase.co/auth/v1/magicLink", () => {
    return HttpResponse.json({ message: "Unable to validate email address: invalid format" }, { status: 400 });
  }),
  // Register
  http.post("https://test.supabase.co/auth/v1/signup", () => {
    const fakeNewAccount = {
      id: "fake_acc_id",
      aud: "authenticated",
      role: "authenticated",
      email: "fake_acc_id@test.com",
      phone: "",
      confirmation_sent_at: Date.now(),
      app_metadata: { provider: "email", providers: ["email"] },
      user_metadata: {},
      identities: [],
      created_at: Date.now(),
      updated_at: Date.now()
    };
    return HttpResponse.json(fakeNewAccount, { status: 200 });
  }),

  http.get("https://test.supabase.co/rest/v1/profiles", ({ request }) => {
    const url = new URL(request.url);
    const name = url.searchParams.get("username")?.replace("eq.", "");
    switch (name) {
      case "supaAwesome":
        return HttpResponse.json(fakenewProfiles[0], { status: 200 });
      case "supaPet":
        return HttpResponse.json(fakenewProfiles[1], { status: 200 });
      default:
        return HttpResponse.json({ message: "ERROR" }, { status: 404 });
    }
  }),

  http.get("https://test.supabase.co/rest/v1/favoritepets", () => {
    const fakeNewFavorites = [
      { id: 1, pet: "1" },
      { id: 2, pet: "2" }
    ];
    return HttpResponse.json(fakeNewFavorites, { status: 200 });
  }),

  http.get("https://test.supabase.co/storage/v1/object/profile/[object%20Object]", () => {
    const myBlob = new Blob(
      [
        `/9j/4AAQSkZJRgABAQEBLAEsAAD/2wBDAAYEBQYFBAYGBQYHBwYIChAKCgkJChQODwwQFxQYGBcUFhYaHSUfGhsjHBYWICwgIyYnKSopGR8tMC0oMCUoKSj/
      2wBDAQcHBwoIChMKChMoGhYaKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCgoKCj/`
      ],
      { type: "image/jpeg" }
    );
    return HttpResponse.json(myBlob, { status: 200 });
  })
);

beforeAll(() => server.listen());
afterAll(() => server.close());
afterEach(() => {
  server.resetHandlers();
});

const SUPABASE_URL = "https://test.supabase.co";
const KEY = "test";
const supabase = createClient(SUPABASE_URL, KEY);

export { server, http, supabase };
