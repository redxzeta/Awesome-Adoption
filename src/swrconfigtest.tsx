import { RenderOptions, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { PetAuthContext, PetTokenType } from "context/TokenContext";
import React, { ReactElement } from "react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-supabase-next";
import { SWRConfig } from "swr";
import { supabase } from "testServer";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <SWRConfig value={{ provider: () => new Map(), dedupingInterval: 0 }}>{children}</SWRConfig>;
};

export const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) => {
  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: AllTheProviders, ...options })
  };
};

type CustomRouteType = {
  route: string;
  name: string;
};

const defaultRoute: CustomRouteType = { route: "/", name: "Testing Page" };

export const customRouterRender = (
  ui: ReactElement,
  routeConfig = defaultRoute,
  options?: Omit<RenderOptions, "wrapper">
) => {
  window.history.pushState({}, routeConfig.name, routeConfig.route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: customProvider, ...options })
  };
};

const customProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <Provider value={supabase}>
      <PetAuthContext.Provider value={initialState}>
        <BrowserRouter>
          <SWRConfig value={{ provider: () => new Map(), dedupingInterval: 0 }}>{children}</SWRConfig>
        </BrowserRouter>
      </PetAuthContext.Provider>
    </Provider>
  );
};
const initialState: PetTokenType = {
  tokenHeaders: "yayeet",
  loading: false,
  errors: false
};
