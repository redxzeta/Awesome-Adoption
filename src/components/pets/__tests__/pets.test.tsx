import { render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

import PetAuthProvider from "../../../context/TokenContext";
import { customRender } from "../../../swrconfigtest";
import { server } from "../../../testServer";
import Pets, { AnimalType } from "../Pets";
import { http, HttpResponse } from "msw";

test("matches animal link snapshot", () => {
  const animalLink = {
    img: "https://via.placeholder.com/150",
    type: "playtpus",
    link: "pets/playtpus"
  };
  const view = render(
    <BrowserRouter>
      <AnimalType type={animalLink.type} img={animalLink.img} link={animalLink.link} />
    </BrowserRouter>
  );
  expect(view).toMatchSnapshot();
});

describe("<Pets/>", () => {
  test("Pets component should render with button", () => {
    render(
      <BrowserRouter>
        <Pets />
      </BrowserRouter>
    );
    const ClickHereButton = screen.getAllByRole("link", {
      name: /Click Here/i
    });
    expect(ClickHereButton[0]).toHaveAttribute("href", "/dog");
    expect(ClickHereButton[1]).toHaveAttribute("href", "/cat");
    expect(ClickHereButton[2]).toHaveAttribute("href", "/bird");
    expect(ClickHereButton[3]).toHaveAttribute("href", "/horse");
    expect(ClickHereButton[4]).toHaveAttribute("href", "/rabbit");
  });

  test("Pet Component list should have attribute href ", () => {
    render(
      <BrowserRouter>
        <Pets />
      </BrowserRouter>
    );
    const petTypeList = screen.getAllByRole("link");
    expect(petTypeList.length).toBe(5);
    expect(petTypeList[0]).toHaveAttribute("href", "/dog");
    expect(petTypeList[1]).toHaveAttribute("href", "/cat");
    expect(petTypeList[2]).toHaveAttribute("href", "/bird");
    expect(petTypeList[3]).toHaveAttribute("href", "/horse");
    expect(petTypeList[4]).toHaveAttribute("href", "/rabbit");
  });

  test("Pet Component should have img with alt attribute", () => {
    render(
      <BrowserRouter>
        <Pets />
      </BrowserRouter>
    );
    const petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(5);
    expect(petImage[0]).toHaveAttribute("alt", "Dog");
    expect(petImage[1]).toHaveAttribute("alt", "Cat");
    expect(petImage[2]).toHaveAttribute("alt", "Bird");
    expect(petImage[3]).toHaveAttribute("alt", "Horse");
    expect(petImage[4]).toHaveAttribute("alt", "Rabbit");
  });

  test("Pets Component should render random pet", async () => {
    customRender(
      <BrowserRouter>
        <PetAuthProvider>
          <Pets />
        </PetAuthProvider>
      </BrowserRouter>
    );

    let petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(5);
    expect(screen.getByRole("status")).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByRole("status")).not.toBeInTheDocument());

    petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(6);
    expect(petImage[5]).toHaveAttribute("alt", "Baby Yoda");
    expect(petImage[5]).toHaveAttribute("src", "babyYoda.medium.jpg");
  });

  test("Pets component should error random pet", async () => {
    server.use(
      http.get("https://api.petfinder.com/v2/animals", () => {
        return HttpResponse.json({ message: "Error" }, { status: 404 });
      })
    );

    customRender(
      <BrowserRouter>
        <PetAuthProvider>
          <Pets />
        </PetAuthProvider>
      </BrowserRouter>
    );

    let petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(5);
    expect(screen.getByRole("status")).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByRole("status")).not.toBeInTheDocument());

    petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(5);
    const errorTitle = screen.getByRole("heading", {
      name: /An Error Occurred/i,
      level: 1
    });

    expect(errorTitle).toBeInTheDocument();
  });

  test("Pets Component should render random pet with placeholder img", async () => {
    server.use(
      http.get("https://api.petfinder.com/v2/animals", () => {
        return HttpResponse.json(
          {
            animal: {
              id: 5,
              name: "Baby Yoda",
              photos: [
                {
                  medium: null
                }
              ]
            }
          },
          { status: 200 }
        );
      })
    );

    customRender(
      <BrowserRouter>
        <PetAuthProvider>
          <Pets />
        </PetAuthProvider>
      </BrowserRouter>
    );

    let petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(5);
    expect(screen.getByRole("status")).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByRole("status")).not.toBeInTheDocument());

    petImage = screen.getAllByRole("img");
    expect(petImage.length).toBe(6);
    expect(petImage[5]).toHaveAttribute("alt", "Baby Yoda");
    expect(petImage[5]).toHaveAttribute("src", "/src/components/pets/placeholder-light.png");
  });
});
