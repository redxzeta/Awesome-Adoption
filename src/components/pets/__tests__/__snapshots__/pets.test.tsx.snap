// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches animal link snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="lg:px-4 px-1 w-full md:1/3 lg:w-1/4 my-4 fadeInUp "
      >
        <div
          aria-label="Card"
          class="card card-bordered image-full"
        >
          <figure>
            <img
              alt="playtpus"
              class="w-full max-h-80 "
              src="https://via.placeholder.com/150"
            />
          </figure>
          <div
            class="card-body"
          >
            <h2
              class="card-title"
            >
              playtpus
               
              Buddy
            </h2>
            <p>
              If a 
              playtpus
               chews shoes whose shoes does he choose?
            </p>
            <div
              class="card-actions justify-end"
            >
              <a
                class="btn btn-primary"
                data-discover="true"
                href="/pets/playtpus"
              >
                Click Here
              </a>
            </div>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="lg:px-4 px-1 w-full md:1/3 lg:w-1/4 my-4 fadeInUp "
    >
      <div
        aria-label="Card"
        class="card card-bordered image-full"
      >
        <figure>
          <img
            alt="playtpus"
            class="w-full max-h-80 "
            src="https://via.placeholder.com/150"
          />
        </figure>
        <div
          class="card-body"
        >
          <h2
            class="card-title"
          >
            playtpus
             
            Buddy
          </h2>
          <p>
            If a 
            playtpus
             chews shoes whose shoes does he choose?
          </p>
          <div
            class="card-actions justify-end"
          >
            <a
              class="btn btn-primary"
              data-discover="true"
              href="/pets/playtpus"
            >
              Click Here
            </a>
          </div>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
